static/js/message_list_view.js

```diff
@@ -1445,4 +1445,103 @@ export class MessageListView {
             message_container.status_message = false;
         }
     }
+
+    /* This function exist for two purposes:
+        * To track the current `sticky_header` which have some different properties
+          like date being always displayed.
+        * Set date on message header corresponding to the message next to the header. */
+    update_sticky_recipient_headers() {
+        const rows_length = this._rows.size;
+        if (!rows_length) {
+            /* No headers are present */
+            return;
+        }
+        /* Intentionally remove sticky headers class here to make calculations simpler. */
+        $(".sticky_header").removeClass("sticky_header");
+        /* visible_top is navbar top position + height for us. */
+        const visible_top = message_viewport.message_viewport_info().visible_top;
+        /* We need date to be properly visible on the header, so partially visible headers
+           who are about to be scrolled out of view are not acceptable. */
+        const header_about_to_be_partially_hidden = visible_top - 1;
+
+        function is_sticky(header) {
+            const header_props = header.getBoundingClientRect();
+            const margin_between_recipient_rows = 10;
+            const header_is_sticky_or_about_to_be_sticky =
+                visible_top + header_props.height + margin_between_recipient_rows;
+            if (header_props.top < header_about_to_be_partially_hidden) {
+                return -1;
+            } else if (header_props.top > header_is_sticky_or_about_to_be_sticky) {
+                return 1;
+            }
+            /* Headers between `header_about_to_be_partially_hidden` and `header_is_sticky_or_about_to_be_sticky`
+               are sticky. If two headers next to each other are completely visible
+               (message header at top has no visible content), we don't mind showing
+               date on any of them. Which header is chosen will depend on which
+               comes first when iterating on the headers. */
+            return 0;
+        }
+
+        const $table = rows.get_table(this.table_name);
+        const $headers = $table.find(".message_header");
+        const iterable_headers = $headers.toArray();
+        let start = 0;
+        let end = iterable_headers.length - 1;
+        let $sticky_header; // This is the first fully visible message header.
+
+        /* Binary search to reach the sticky header */
+        while (start <= end) {
+            const mid = Math.floor((start + end) / 2);
+            const header = iterable_headers[mid];
+            const diff = is_sticky(header);
+            if (diff === 0) {
+                $sticky_header = $(header);
+                break;
+            } else if (diff === 1) {
+                end = mid - 1;
+            } else {
+                start = mid + 1;
+            }
+        }
```

Comment:
Binary search is one of those algorithms that, while simple, is really easy to get wrong around boundary conditions (lists with 0-1 elements, etc). Can you help me convince myself this is correct?

Reply:
For the corner cases:
* 0: Not possible since we test for it at the start of the function:
```js
        if (!rows_length) {
            /* No headers are present */
            return;
        }
```
* 1: In this case: header can be either sticky if user scrolls a very long message or the header is not sticky and completely visible since we are at the top of message feed. If the header is sticky, we set `$sticky_header` and proceed as normal, if not, we exit the loop with `$sticky_header` as `undefined` which is also handled just after this code.